### PR TITLE
Improve performance. Add brunch stats.

### PR DESCRIPTION
**Pull request by kuhar**

### CODE REVIEW

File: include/sea_dsa/Graph.hh [hunk 1]
```diff
@@ -443,7 +443,7 @@ private:
 
 public:
   typedef Graph::Set Set;
-  typedef boost::container::flat_map<unsigned, Set> accessed_types_type;
+  typedef llvm::DenseMap<unsigned, Set> accessed_types_type;
```
Comment (kuhar) [inline include/sea_dsa/Graph.hh]: this automagically shaved-off ~10% of running time in addAccessedTypes.

File: src/Graph.cc [hunk 2]
```diff
@@ -121,56 +121,86 @@ bool sea_dsa::Node::hasAccessedType(unsigned o) const {
...
 }
 
-void sea_dsa::Node::addAccessedType(unsigned o, llvm::Type *t) {
-  if (isOffsetCollapsed())
-    return;
-  Offset offset(*this, o);
-  growSize(offset, t);
-  if (isOffsetCollapsed())
-    return;
+void sea_dsa::Node::addAccessedType(unsigned off, llvm::Type *type) {
```
Comment (kuhar) [inline src/Graph.cc]: I noticed addAccessedTypes was a hotspot in SMC and modified it to take ~30% fewer cycles

File: include/sea_dsa/Graph.hh (hunk 1 above)
Comment (agurfinkel) [reply include/sea_dsa/Graph.hh]: Hmm... flag for more investigation. Either we are not using `flat_map` correctly, or `accessed_types_type` is fairly big

File: include/sea_dsa/Graph.hh [hunk 3]
```diff
@@ -679,7 +680,9 @@ public:
 
   const Set getAccessedType(unsigned o) const {
     Offset offset(*this, o);
-    return m_accessedTypes.at(offset.getNumericOffset());
+    auto it = m_accessedTypes.find(offset.getNumericOffset());
```
Comment (agurfinkel) [inline include/sea_dsa/Graph.hh]: this seems unnecessary. There should be an assert in `at()` already

File: include/sea_dsa/support/Brunch.hh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+template <typename T>
+void PrintBrunchProgress(llvm::Twine key, const T &value, const T &max = "") {
+  llvm::errs() << "BRUNCH_PROGRESS " << key << " " << value << " / " << max
+               << "\n";
+}
+
+long GetCurrentMemoryUsageKb();
+
+class BrunchTimer {
```
Comment (agurfinkel) [inline include/sea_dsa/support/Brunch.hh]: We already have a stopwatch and scoped timers. Why is this necessary? If you don't like how we measure time, change it there.

File: src/DsaBottomUp.cc [hunk 5]
```diff
@@ -138,10 +136,22 @@ bool BottomUpAnalysis::runOnModule(Module &M, GraphMap &graphs) {
...
+
   LocalAnalysis la(m_dl, m_tli, m_allocInfo);
+
+  const size_t totalFunctions = M.getFunctionList().size();
+  size_t functionsProcessed = 0;
+
   for (auto it = scc_begin(&m_cg); !it.isAtEnd(); ++it) {
     auto &scc = *it;
 
+    SEA_DSA_BRUNCH_PROGRESS("BU_FUNCTIONS_PROCESSED", functionsProcessed,
```
Comment (agurfinkel) [inline src/DsaBottomUp.cc]: You should use Stats for this instead of creating extra local variables

File: src/DsaTopDown.cc [hunk 6]
```diff
@@ -99,9 +98,18 @@ bool TopDownAnalysis::runOnModule(Module &M, GraphMap &graphs) {
...
+  size_t functionsProcessed = 0;
+  BrunchTimer tdTimer("TD");
+
   for (auto it = postorder_scc.rbegin(), et = postorder_scc.rend(); it != et;
        ++it) {
     auto &scc = *it;
+
+    SEA_DSA_BRUNCH_PROGRESS("TD_FUNCTIONS_PROCESSED", functionsProcessed,
+                            totalFunctions);
+    functionsProcessed += scc.size();
```
Comment (agurfinkel) [inline src/DsaTopDown.cc]: Ouch. the stats all over the place is very annoying :(

File: include/sea_dsa/support/Brunch.hh (hunk 4 above)
Comment (kuhar) [reply include/sea_dsa/support/Brunch.hh]: it's in seahorn, not in sea_dsa

File: src/DsaBottomUp.cc (hunk 5 above)
Comment (kuhar) [reply src/DsaBottomUp.cc]: same as above -- it's in seahorn

File: src/DsaTopDown.cc (hunk 6 above)
Comment (kuhar) [reply src/DsaTopDown.cc]: it's enabled by a separate flag, `--log=brunch-progress`, so shouldn't bother anyone.
we can also print it more rarely, say every 1% of progress or something

File: include/sea_dsa/support/Brunch.hh (hunk 4 above)
Comment (agurfinkel) [reply include/sea_dsa/support/Brunch.hh]: at the moment we just copy code instead of creating new and different one

File: src/DsaTopDown.cc (hunk 6 above)
Comment (agurfinkel) [reply src/DsaTopDown.cc]: I mean code for collecting and managing stats is all over the place in the code. Of course I see how this is controlled at runtime. I'm annoyed by code bloat not performance.

File: include/sea_dsa/Graph.hh (hunk 3 above)
Comment (kuhar) [reply include/sea_dsa/Graph.hh]: there's no at in llvm::DenseMap

Comment (agurfinkel) [reply include/sea_dsa/Graph.hh]: there is `lookup`

File: src/DsaTopDown.cc (hunk 6 above)
Comment (agurfinkel) [reply src/DsaTopDown.cc]: Also in the current code, `BRUNCH_STAT` is restricted to `Stats` only and the magic string is not spread out throughout the code (I hope ;) )

**Comment by kuhar** [comment]
@agurfinkel I committed performance and correctness fixes and left the stats and timers on this branch.

**Comment by kuhar** [comment]
No sure why it didn't pick up my updated commit message, though...

**Comment by kuhar** [comment]
@agurfinkel 
> I dont see how this takes care of goobals that are passed but not through arguments. They would be the bigger issue

The current version is also unsound... I'm working on a better way to handle that.

But I don't see how this optimisation would apply to global not passed by arguments. Thinking about it, it should be possible to do something very similar for return values during bottom-up.

**Comment by kuhar** [comment]
OK, I think that transitive allocation site pruning can be done in the Cloner if we can build a call-site-specific local memory graph for the cloned actuals. I don't know if this can be easily done now with a simple local reasoning (e.g., BasicAA).
@agurfinkel what's your idea for that?
